### PR TITLE
Add complex PSD cone

### LINKED ISSUE #193: Support for the complex PSD cone
Are you interested in adding direct support for the complex PSD cone? It's much faster than solving a complex problem via the embedding in the real PSD cone.

Currently the only solver that has it is [Hypatia](https://github.com/jump-dev/Hypatia.jl), so it'd be great to have a first-order solver that supports it as well.

### PR DESCRIPTION
**Pull request by araujoms**
Closes #193.

I changed the typedefs of `PsdConeTriangle` and `DensePsdConeTriangle` to accept an extra parameter, which determines whether the cone is real or complex. I generalized all relevant code to handle real or complex matrices. Also I added the new cones to the list of supported by MOI.

I did some benchmarking for a simple test problem that computes the minimal eigenvalue of a random complex Hermitian matrix. The orange curve is the times for the native complex version as a function of the dimension, whereas the blue curve refers to the the same problem mapped onto real Hermitian matrices via the usual technique `c -> [real(c) imag(c); -imag(c) real(c)]`:
complex_versus_real
As you can see, the difference is brutal. The code I used for benchmark is the one below:

```julia
using LinearAlgebra
using COSMO
import Random

function random_herm(::Type{T}, d) where {T}
    a = randn(T, d,d)
    return Hermitian(a + a')
end

function benchmark(::Type{T}, d) where {T}
    Random.seed!(2988)
    c = random_herm(complex(T), d)
    if T <: Real
        real_c = Hermitian([real(c) imag(c); -imag(c) real(c)])
        return mineig_raw(real_c)
    else
        return mineig_raw(c)        
    end
end

tri(d) = div(d*(d+1), 2)
    
function mineig_raw(c::AbstractMatrix{R}) where {R}
    d = size(c, 1)
    T = real(R)
    vec_dim = R <: Complex ? d^2 : tri(d)    

    model = COSMO.Model{T}()
    vec_c = zeros(T, vec_dim)
    COSMO.extract_upper_triangle!(c, vec_c, sqrt(T(2)))

    P = zeros(T, vec_dim, vec_dim)

    id_vec = zeros(T, vec_dim)
    diagonal_indices = tri.(1:d)
    id_vec[diagonal_indices] .= 1

    cs1 = COSMO.Constraint(id_vec', -T(1), COSMO.ZeroSet)
    cs2 = COSMO.Constraint(Matrix(T(1)*I(vec_dim)), zeros(T,vec_dim), COSMO.PsdConeTriangle{T, R}(vec_dim))
    constraints = [cs1; cs2]

    assemble!(model, P, vec_c, constraints, settings = COSMO.Settings{T}(verbose = false, eps_abs = 1e-3, eps_rel = 1e-3, eps_prim_inf = 1e-3, eps_dual_inf = 1e-3))
    result = COSMO.optimize!(model)
#    display(minimum(eigvals(c)))
#    display(result.obj_val)
    return result.times.solver_time
end
```
I've also checked that it works with JuMP with the following code:
```julia
using JuMP
using LinearAlgebra
using COSMO

function mineig(c::AbstractMatrix{R}) where {R}
    d = size(c, 1)
    T = real(R)

    model = GenericModel{T}(COSMO.Optimizer{T})
    if R <: Complex
        @variable(model, rho[1:d,1:d] in HermitianPSDCone())
    else
        @variable(model, rho[1:d,1:d] in PSDCone())
    end
    @constraint(model, tr(rho) == 1)
    @objective(model, Min, real(dot(c,rho)))
    set_attribute(model, "verbose", true)   
    optimize!(model)
    
    display(minimum(eigvals(c)))
    return objective_value(model)
end
```

I noticed two problems: COSMO often fails to converge, hitting the iteration limit (the peaks of the graph above), and doesn't work for Float32. Both problems were there before my changes, though.

## ☑️ PR checklist

- [x] I have linked to the issue nr that this PR aims to resolve.
- [x] If this PR is not yet ready to be merged, I have set it to "Draft".
- [x] I have added tests for new features or the bug that this PR aims to fix.
- [x] I have provided docstrings for new functions.
- [x] I updated the repository documentation if these code changes touch the interface.

### CODE REVIEW

**Comment by migarstka** [comment]
Great. Thank you @araujoms for your effort on this. 

I am a bit occupied with my day job this week, but I try to review either one evening or on the weekend.

**Comment by araujoms** [comment]
Sure, I've added a unit test using the native interface. I could also add one using JuMP, but there doesn't seem to be any.

**Comment by migarstka** [comment]
I can see one test failure on Julia 1.6:

```
     Testing Running tests...
Complex PSD Cone: Test Failed at /home/runner/work/COSMO.jl/COSMO.jl/test/UnitTests/least_eigenvalue.jl:35
  Expression: ≈(mineig_raw(X), minimum(eigvals(X)), atol = 0.0001, rtol = 0.0001)
   Evaluated: -0.6468784469158599 ≈ -1.9271575272948176 (atol=0.0001, rtol=0.0001)
```

**Comment by araujoms** [comment]
Well you could drop support for 1.6. It runs fine on 1.9, and 1.10 will soon be the new LTS anyway.

**Comment by araujoms** [comment]
As it turns out Julia 1.6 is innocent, I just ran the tests locally, and they went fine. I guess it was just a random instability, I made the test deterministic to avoid that.

File: test/run_cosmo_tests.jl [hunk 1]
```diff
@@ -4,14 +4,15 @@ rng = Random.MersenneTwister(12345)
 
 
 
-include("./UnitTests/COSMOTestUtils.jl")
+#include("./UnitTests/COSMOTestUtils.jl")
```
Comment (migarstka) [inline test/run_cosmo_tests.jl:7]: any particular reason why you comment this file out?

**Comment by migarstka** [comment]
It looks good so far. Maybe we should point potential users at this feature in the docs. The section on the PsdConeTriangle is here:

https://github.com/oxfordcontrol/COSMO.jl/blob/c3a04f0377982ecc180f07f33dfa2ad8097d7a03/docs/src/getting_started.md?plain=1#L35

Either by adding a note under the constraints. Or by adding a sentence to the PsdConeTriangle that one can use complex numbers as well (and linking to the test)

File: test/run_cosmo_tests.jl (hunk 1 above)
Comment (araujoms) [reply test/run_cosmo_tests.jl:7]: It is already included in `runtests.jl`. The duplicate inclusion was generating a lot of warnings about function redefinitions.

**Comment by araujoms** [comment]
Done. I added another line to the list of cones in Getting Started, and gave more detail in the docstring.

**Comment by migarstka** [comment]
Thanks a lot

**Comment by araujoms** [comment]
My pleasure. Thanks for merging.

**Pull request merged**